docs/devel/writing-good-e2e-tests.md

```diff
@@ -0,0 +1,168 @@
+useless in the harsh reality of a CI environment.
+ 
+
+1. **Debuggability**: if your test fails, it should provide as
+detailed as possible reasons for the failure in it's output. "Timeout"
+is not a useful error message. "Timed out after 60 seconds waiting for
+pod xxx to enter running state, still in pending state" is much more
+useful to someone trying to figure out why your test failed and what
+to do about it.  It goes without saying that overly verbose logging,
+particularly of non-error conditions, can make it unnecessarily
+difficult to figure out whether a test failed and if so why?  So don't
+log lots of irrelevant stuff either.
+1. **Ability to run in non-dedicated test clusters:** To reduce
+end-to-end delay and improve resource utilization when running e2e
+tests, we try, where possible, to run large numbers of tests in
+parallel against the same test cluster.  This means that:
+    1. you should avoid making any assumption (implicit or explicit)
+that your test is the only thing running against the cluster.  For
+example, making the assumption that your test can run a pod on every
+node in a cluster is not a safe assumption, as some other tests,
+running at the same time as yours, might have saturated one or more
+nodes in the cluster.  Similarly, running a pod in the system
+namespace, and assuming that that will increase the count of pods in
+the system namespace by one is not safe, as some other test might be
+creating or deleting pods in the system namespace at the same time as
+your test.  If you do legitimately need to write a test like that,
+make sure to label it "SERIAL" so that it's easy to identify, and not
+run in parallel with any other tests.
+    1. You should avoid doing things to the cluster that make it
+difficult for other tests to reliably do what they're trying to do, at
+the same time.  For example, rebooting nodes, disconnecting network
+interfaces, or upgrading cluster software as part of your test is
+likely to violate the assumptions that other tests might have made
+about a reasonably stable cluster environment.  If you need to write
+such tests, please label them as "DISRUPTIVE" so that it's easy to
+identify them, and not run them in parallel with other tests.
+1. **Speed of execution:** We have hundreds of e2e tests, some of
+which we run in serial, one after the other, in some cases.  If each
+test takes just a few minutes to run, that very quickly ads up to
+many, many hours of total execution time.  We try to keep such total
+execution time down to a few tens of minutes at most.  Therefore, try
+(very hard) to keep the execution time of your individual tests below
+2 minutes, ideally shorter than that.  Concretely, adding
+inappropriately long 'sleep' statements or other gratuitous waits to
+tests is a killer.  If under normal circumstances your pod enters the
+running state within 10 seconds, and 99.9% of the time within 30
+seconds, it would be gratuitous to wait 5 minutes for this to happen.
```

Comment:
@brendandburns I know this has been contentious in the past.  Want to add your 2¢?